File: linear_scan/linearscan.py
import matplotlib.pyplot as plt

def expireOldIntervals(interval, active, registers):
    for i in active.copy():
        if i[1] <= interval[0]:
            active.remove(i)
            for j, reg in enumerate(registers):
                if reg == i:
                    registers[j] = (-1, -1)

def displayRegisters(registers, dick):
    for i, reg in enumerate(registers):
        variable = next((key for key, value in dick.items() if value == (reg[0], reg[1])), 'NONE')
        print(f"Register {i+1} contains variable {variable} (live interval: {dick[variable][0]}-{dick[variable][1]})")
    print()

def linearScanRegisterAllocation(LiveIntervals, registers, dick):
    active = set()

    for interval in LiveIntervals:
        expireOldIntervals(interval, active, registers)

        if len(active) == R:
            spillAtInterval(interval, active, registers, dick)
        else:
            for i, reg in enumerate(registers):
                if reg == (-1, -1):
                    registers[i] = interval
                    active.add(interval)
                    break

        displayRegisters(registers, dick)

def spillAtInterval(interval, active, registers, dick):
    max_end = interval[1]
    flag = False
    spill_interval = interval

    for i in active:
        if i[1] > max_end:
            max_end = i[1]
            spill_interval = i
            flag = True

    if flag:
        active.remove(spill_interval)
        variable = next((key for key, value in dick.items() if value == (spill_interval[0], spill_interval[1])), 'NONE')
        print(f"Spilled variable {variable} to memory")
        index = -1
        for j, reg in enumerate(registers):
            if reg == spill_interval:
                registers[j] = (-1, -1)
                index = j
                break
        active.add(interval)
        registers[index] = interval
    else:
        variable = next((key for key, value in dick.items() if value == (spill_interval[0], spill_interval[1])), 'NONE')
        print(f"Spilled variable {variable} to memory")

if __name__ == "__main__":
    LiveIntervals = {(1, 3), (2, 5), (3, 10), (4, 8), (5, 7)}
    #make a dictionary of intervals mapping them to variables
    dick={'a':(1, 3), 'b':(2, 5), 'c':(3, 10), 'd':(4, 8), 'e':(5, 7),'NONE':(-1,-1)}
    R = 2
    registers = [(-1, -1)] * R

    for interval in LiveIntervals:
        print(interval[0], interval[1])

    linearScanRegisterAllocation(LiveIntervals, registers, dick)

File: linearscan.cpp
#include <iostream>
#include <set>
#include <vector>
using namespace std;
#define R 3
// Linear Scan Register Allocation

void expireOldIntervals(pair<int,int> interval, set <pair <int, int>> &active,vector<pair<int,int>>& registers);
void displayRegisters(vector<pair<int,int>>& registers);
void linearScanRegisterAllocation(set <pair <int, int>> &LiveIntervals, vector<pair<int,int>>& registers);
void spillAtInterval(pair <int, int> interval, set <pair <int, int>> &active,vector<pair<int,int>>& registers);


void displayRegisters(vector<pair<int,int>>& registers){
    for(int i=0;i<registers.size();i++){
        cout<<"Register "<<i+1<<" contains "<<registers[i].first<<" "<<registers[i].second<<endl;
    }
    cout<<endl;
}

void linearScanRegisterAllocation(set <pair <int, int>> &LiveIntervals, vector<pair<int,int>>& registers){
    set <pair <int, int>> active;
    
    for(auto it = LiveIntervals.begin(); it != LiveIntervals.end(); it++){
        expireOldIntervals(*it,active,registers);

        if(active.size() == R){
            spillAtInterval(*it, active, registers);
        }
        else{
            for(int i=0;i<registers.size();i++){
                if(registers[i] == make_pair(-1,-1)){
                    registers[i] = *it;
                    active.insert(*it);
                    break;
                }
            }
        }
        displayRegisters(registers);
    }
    
}

void expireOldIntervals(pair<int,int> interval, set <pair <int, int>> &active,vector<pair<int,int>>& registers){
    for(auto it = active.begin(); it != active.end(); it++){
        pair<int,int> temp = *it;
        if(it->second <= interval.first){
            active.erase(*it);
            for(int i=0;i<registers.size();i++){
                if(registers[i] == temp){
                    registers[i] = make_pair(-1,-1);
                }
            }
        }
    }
}


void spillAtInterval(pair <int, int> interval, set <pair <int, int>> &active,vector<pair<int,int>>& registers){

    int max = interval.second;

    bool flag = false;

    pair<int,int> spillInterval = interval;
    for(auto it = active.begin(); it != active.end(); it++){
        if(it->second > max){
            max = it->second;
            spillInterval = *it;
            flag = true;
        }
    }

    if(flag){
        
        active.erase(spillInterval);
        cout<<"Spilled "<<spillInterval.first<<" "<<spillInterval.second<<"to memory"<<endl;
        int index=-1;
        for(int i=0;i<registers.size();i++){
            if(registers[i] == spillInterval){
                registers[i] = make_pair(-1,-1);
                index = i;
                break;
            }
        }
        active.insert(interval);
        registers[index] = interval;
    }else{
        cout<<"Spilled "<<interval.first<<" "<<interval.second<<" to memory"<<endl;
    }

}

    
int main(){
    
    set <pair <int, int>> LiveIntervals = {{1,3}, {2,5}, {3,10}, {4,8}, {5,7}};
    vector<pair<int,int>> registers(R, make_pair(-1,-1));
    //iterate through the set

    for (auto it = LiveIntervals.begin(); it != LiveIntervals.end(); it++){
        cout<< it -> first <<" "<< it -> second << endl;
    }

    linearScanRegisterAllocation(LiveIntervals, registers);


    
  
  return 0;
}